sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java

```diff
@@ -0,0 +1,231 @@
+  }
+
+  private void addBreadcrumb(final @NonNull Activity activity, final @NotNull String state) {
+    if (options.isEnableActivityLifecycleBreadcrumbs()) {
+      final Breadcrumb breadcrumb = new Breadcrumb();
+      breadcrumb.setType("navigation");
+      breadcrumb.setData("state", state);
+      breadcrumb.setData("screen", getActivityName(activity));
+      breadcrumb.setCategory("ui.lifecycle");
+      breadcrumb.setLevel(SentryLevel.INFO);
+      hub.addBreadcrumb(breadcrumb);
+    }
+  }
+
+  private @NotNull String getActivityName(final @NonNull Activity activity) {
+    return activity.getClass().getSimpleName();
+  }
+
+  private void stopPreviousTransactions() {
+    for (final Map.Entry<Activity, ITransaction> entry :
+        activitiesWithOngoingTransactions.entrySet()) {
+      final ITransaction transaction = entry.getValue();
+      finishTransaction(transaction);
+    }
+  }
+
+  private void startTracing(final @NonNull Activity activity) {
+    if (performanceEnabled && !isRunningTransaction(activity)) {
+      // as we allow a single transaction running on the bound Scope, we finish the previous ones
+      stopPreviousTransactions();
+
+      final ITransaction transaction =
+          hub.startTransaction(getActivityName(activity), "navigation");
+
+      // lets bind to the scope so other integrations can pick it up
+      hub.configureScope(
+          scope -> {
+            applyScope(scope, transaction);
+          });
+
+      activitiesWithOngoingTransactions.put(activity, transaction);
+    }
+  }
+
+  @VisibleForTesting
+  void applyScope(final @NotNull Scope scope, final @NotNull ITransaction transaction) {
+    scope.withTransaction(
+        scopeTransaction -> {
+          // we'd not like to overwrite existent transactions bound to the Scope
+          // manually.
+          if (scopeTransaction == null) {
+            scope.setTransaction(transaction);
+          } else {
+            options
+                .getLogger()
+                .log(
+                    SentryLevel.DEBUG,
+                    "Transaction '%s' won't be bound to the Scope since there's one already in there.",
+                    transaction.getName());
+          }
+        });
+  }
+
+  private boolean isRunningTransaction(final @NonNull Activity activity) {
+    return activitiesWithOngoingTransactions.containsKey(activity);
+  }
+
+  private void stopTracing(final @NonNull Activity activity, final boolean shouldFinishTracing) {
+    if (performanceEnabled && shouldFinishTracing) {
+      final ITransaction transaction = activitiesWithOngoingTransactions.get(activity);
+      finishTransaction(transaction);
+    }
+  }
+
+  private void finishTransaction(final @Nullable ITransaction transaction) {
+    if (transaction != null) {
+      SpanStatus status = transaction.getStatus();
+      // status might be set by other integrations, let's not overwrite it
+      if (status == null) {
```

Comment:
If status can be set in other threads perhaps status could be also an atomic reference and `compareAndSet` could be used.

Reply:
that's a good point